### PR TITLE
Integration of property based testing into Bitcoin Core

### PR DESCRIPTION
**Pull request by Christewart**
This PR is a subset of the changes in #8469. It's meant to be easier to review. This PR contains all of the build instructions needed for travis to pass. It includes one property call `key_properties.cpp` along with a generator file called `crypto_gen.{h,cpp}`.

### CODE REVIEW

**Comment by practicalswift** [comment]
Strong concept ACK

Very nice work!

**Comment by randolf** [comment]
@Christewart I noticed that ``depends/packages/rapidcheck.mk`` references MarcoFalke:

``$(package)_download_path=https://github.com/MarcoFalke/rapidcheck/archive``

Do you think it might be better to reference directly from "bitcoin" instead?  If so, should MarcoFalke's "rapidcheck" project be added to the "bitcoin" project to facilitate this?

**Comment by randolf** [comment]
@Christewart In ``src/test/key_properties.cpp`` there are new references to the BOOST libraries.  As I understand it, one of the goals is to move away from BOOST, so I'm curious:  Is it possible to accomplish the same goals in this code without the additional reliance on BOOST?

**Comment by MarcoFalke** [comment]
We don't plan to get rid of the boost unit test framework any time soon.

**Comment by Christewart** [comment]
@randolf I'm not sure if it is too much of a concern right now who hosts the package. If we want to move it under the bitcoin project in the future i have no problem with that. Marco was kind enough to help me out with the setup which is why his repo is currently referenced.

**Comment by MarcoFalke** [comment]
@randolf Ideally it should reference the proper upstream, but since they don't have any releases that might not work.

**Comment by laanwj** [comment]
> We don't plan to get rid of the boost unit test framework any time soon.

Indeed. @randolf In general: please don't comment on boost usage of PRs in reviews if it concerns libraries that are already used in other parts of our code (such as boost test), only if it concerns using new boost libaries (such as lexical cast and whatnot.)

File: src/test/gen/crypto_gen.cpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#include "test/gen/crypto_gen.h"
```
Comment (fanquake) [inline src/test/gen/crypto_gen.cpp]: This and the .h are missing a copyright header:
```
// Copyright (c) 2018 The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```

File: src/test/key_properties.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
+// Copyright (c) 2012-2016 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
Comment (fanquake) [inline src/test/key_properties.cpp]: This can just be:
```
// Copyright (c) 2018 The Bitcoin Core developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```

File: depends/packages/rapidcheck.mk [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package=rapidcheck
+$(package)_version=10fc0cb
+$(package)_download_path=https://github.com/MarcoFalke/rapidcheck/archive
+$(package)_file_name=$(package)-$($(package)_version).tar.gz
+$(package)_sha256_hash=9640926223c00af45bce4c7df8b756b5458a89b2ba74cfe3e404467f13ce26df
+
+define $(package)_config_cmds
+  cmake -DCMAKE_POSITION_INDEPENDENT_CODE:BOOL=true .
```
Comment (fanquake) [inline depends/packages/rapidcheck.mk]: Looks like we'll need some documentation about installing/having cmake available
```
Extracting rapidcheck...
/Users/xxx/GitHub/bitcoin/depends/sources/rapidcheck-10fc0cb.tar.gz: OK
Preprocessing rapidcheck...
Configuring rapidcheck...
/bin/sh: cmake: command not found
```

**Comment by fanquake** [comment]
Testing https://github.com/bitcoin/bitcoin/pull/12775/commits/900a441eeacd9f936a2495e878174f9aeead66d2 on macOS 10.13.4:

After installing ```cmake``` I could successfully build rapidcheck in depends:
```
bash-3.2$ make NO_QT=1 RAPIDCHECK=1
Configuring rapidcheck...
-- The C compiler identification is AppleClang 9.1.0.9020039
-- The CXX compiler identification is AppleClang 9.1.0.9020039
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/xxx/GitHub/bitcoin/depends/work/build/x86_64-apple-darwin17.5.0/rapidcheck/10fc0cb-3719dbaacbb
Building rapidcheck...
Scanning dependencies of target rapidcheck
[  2%] Building CXX object CMakeFiles/rapidcheck.dir/src/BeforeMinimalTestCase.cpp.o
[  5%] Building CXX object CMakeFiles/rapidcheck.dir/src/Check.cpp.o
[  8%] Building CXX object CMakeFiles/rapidcheck.dir/src/Classify.cpp.o
[ 11%] Building CXX object CMakeFiles/rapidcheck.dir/src/GenerationFailure.cpp.o
[ 14%] Building CXX object CMakeFiles/rapidcheck.dir/src/Log.cpp.o
[ 17%] Building CXX object CMakeFiles/rapidcheck.dir/src/Random.cpp.o
[ 20%] Building CXX object CMakeFiles/rapidcheck.dir/src/Show.cpp.o
[ 22%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Any.cpp.o
[ 25%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Assertions.cpp.o
[ 28%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Base64.cpp.o
[ 31%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Configuration.cpp.o
[ 34%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/DefaultTestListener.cpp.o
[ 37%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/FrequencyMap.cpp.o
[ 40%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/ImplicitParam.cpp.o
[ 42%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/LogTestListener.cpp.o
[ 45%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/MapParser.cpp.o
[ 48%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/MulticastTestListener.cpp.o
[ 51%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/ParseException.cpp.o
[ 54%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Platform.cpp.o
[ 57%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Property.cpp.o
[ 60%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/PropertyContext.cpp.o
[ 62%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/ReproduceListener.cpp.o
[ 65%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Results.cpp.o
[ 68%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Serialization.cpp.o
[ 71%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/StringSerialization.cpp.o
[ 74%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/TestMetadata.cpp.o
[ 77%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/TestParams.cpp.o
[ 80%] Building CXX object CMakeFiles/rapidcheck.dir/src/detail/Testing.cpp.o
[ 82%] Building CXX object CMakeFiles/rapidcheck.dir/src/gen/Numeric.cpp.o
[ 85%] Building CXX object CMakeFiles/rapidcheck.dir/src/gen/Text.cpp.o
[ 88%] Building CXX object CMakeFiles/rapidcheck.dir/src/gen/detail/ExecHandler.cpp.o
[ 91%] Building CXX object CMakeFiles/rapidcheck.dir/src/gen/detail/GenerationHandler.cpp.o
[ 94%] Building CXX object CMakeFiles/rapidcheck.dir/src/gen/detail/Recipe.cpp.o
[ 97%] Building CXX object CMakeFiles/rapidcheck.dir/src/gen/detail/ScaleInteger.cpp.o
[100%] Linking CXX static library librapidcheck.a
[100%] Built target rapidcheck
Staging rapidcheck...
Postprocessing rapidcheck...
Caching rapidcheck...
```
Did a ```./configure --prefix=`pwd`/depends/x86_64-apple-darwin17.5.0```:
```
checking rapidcheck.h usability... yes
checking rapidcheck.h presence... yes
checking for rapidcheck.h... yes
```

You could add rapidcheck under the "Options used to compile and link:" at the end of ```./configure```

```make check -j6``` failed with:
```
Running tests: from test/gen/crypto_gen.cpp
Missing an argument value for the parameter run_test in the argument 

Parameter: run_test
 Filters, which test units to include or exclude from test module execution.
 Command line formats:
   --run_test=<test unit filter>
   -t <test unit filter>
 Environment variable: BOOST_TEST_RUN_FILTERS

For detailed help on Boost.Test parameters use:
  test_bitcoin --help
or
  test_bitcoin --help=<parameter name>
make[3]: *** [test/gen/crypto_gen.cpp.test] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [check-am] Error 2
make[1]: *** [check-recursive] Error 1
make: *** [check-recursive] Error 1
```

**Comment by MarcoFalke** [comment]
Could you set `RAPIDCHECK` for the depends build and `--with-rapidcheck` for the configure, such that travis runs it?

**Comment by Christewart** [comment]
@fanquake I fixed the issue with `make check`. I was accidentally classifying generator files as actual unit test files

**Comment by Christewart** [comment]
one build is still failing from a timeout issue it appears? I don't think it is related to this pr?

**Comment by laanwj** [comment]
Probably a good time to move this forward and finally get it in now that 0.17 is branched off.

**Comment by Christewart** [comment]
@laanwj Rebased

File: configure.ac [hunk 4]
```diff
@@ -1134,6 +1140,22 @@ AC_CHECK_DECLS([EVP_MD_CTX_new],,,[AC_INCLUDES_DEFAULT
...
+elif test "x$use_rapidcheck" != xno; then
+    enable_property_tests=yes
+fi
+
+RAPIDCHECK_LIBS=
+if test "x$enable_property_tests" = xyes; then
+   RAPIDCHECK_LIBS=-lrapidcheck
+fi
+AC_SUBST(RAPIDCHECK_LIBS)
+AM_CONDITIONAL([ENABLE_PROPERTY_TESTS], [test x$enable_property_tests = xyes])
```
Comment (laanwj) [inline configure.ac]: I think something should be logged to the output depending on whether rapidcheck is used or not, currently there is nothing in configure output that suggests this

File: depends/packages/rapidcheck.mk [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+endef
+
+define $(package)_build_cmds
+  $(MAKE) && \
+  mkdir -p $($(package)_staging_dir)$(host_prefix)/include && \
```
Comment (laanwj) [inline depends/packages/rapidcheck.mk]: indeed, this is necessary because rapidcheck has no `make install`

**Comment by laanwj** [comment]
I'm surprised that debian/ubuntu doesn't package `rapidcheck`, the only way to get it is to build it from source.

**Comment by laanwj** [comment]
> This PR contains all of the build instructions needed for travis to pass. 

it is not enabled in travis.yml though, is that intentional?

File: depends/packages/rapidcheck.mk (hunk 5 above)
Comment (Christewart) [reply depends/packages/rapidcheck.mk]: This is an open PR on the repo

https://github.com/emil-e/rapidcheck/pull/152

**Comment by Christewart** [comment]
>it is not enabled in travis.yml though, is that intentional?

Frankly I haven't got that far. It's not intentional, just extra work that I don't have time to do ATM :-).

**Comment by fanquake** [comment]
@*Christewart If you’re currently **short on time, I can take this over.*

**Comment by laanwj** [comment]
> Frankly I haven't got that far. It's not intentional, just extra work that I don't have time to do ATM :-).

It's fine, doesn't have to be done in this PR.
It's just that your opening post mentions "This PR contains all of the build instructions needed for travis to pass."

File: src/test/gen/crypto_gen.cpp [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+#include <key.h>
+
+#include <rapidcheck/gen/Arbitrary.h>
+#include <rapidcheck/Gen.h>
+#include <rapidcheck/gen/Predicate.h>
+#include <rapidcheck/gen/Container.h>
+
+/** Generates 1 to 20 keys for OP_CHECKMULTISIG */
+rc::Gen<std::vector<CKey>> MultisigKeys()
```
Comment (practicalswift) [inline src/test/gen/crypto_gen.cpp]: `MultisigKeys()` seems unused? Is that intentional?

**Comment by pawan17kumar** [comment]
Please fix it

File: src/test/gen/crypto_gen.cpp (hunk 6 above)
Comment (Christewart) [reply src/test/gen/crypto_gen.cpp]: I don't think it is used on this PR, but I believe it is used with the remaining test cases in #8469. Basically this PR split #8469 in half -- there is still a lot of test cases in that PR. Basically once this PR is merged i'm going to open another one with those test cases

File: src/Makefile.test.include [hunk 7]
```diff
@@ -8,6 +8,7 @@ TEST_SRCDIR = test
 TEST_BINARY=test/test_bitcoin$(EXEEXT)
 
 JSON_TEST_FILES = \
+  test/data/script_tests.json \
```
Comment (MarcoFalke) [inline src/Makefile.test.include]: Why is this change required?

File: depends/packages/rapidcheck.mk [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  cp -a include/* $($(package)_staging_dir)$(host_prefix)/include/ && \
+  cp -a extras/boost_test/include/rapidcheck/* $($(package)_staging_dir)$(host_prefix)/include/rapidcheck/ && \
+  mkdir -p $($(package)_staging_dir)$(host_prefix)/lib && \
+  cp -a librapidcheck.a $($(package)_staging_dir)$(host_prefix)/lib/
```
Comment (MarcoFalke) [inline depends/packages/rapidcheck.mk]: Might want to update to the latest rapidcheck and use the make install target?

File: depends/packages/rapidcheck.mk (hunk 5 above)
Comment (Empact) [reply depends/packages/rapidcheck.mk]: Now that the PR is merged, should this be switched over?

File: src/test/gen/crypto_gen.cpp (hunk 6 above)
Comment (Empact) [reply src/test/gen/crypto_gen.cpp]: I think we should be more careful about adding unused code like this under the expectation that it will be used in the future. YAGNI + concerns about maintaining an informative record in the code history.
